src/engine/BMInterface.php

```diff
@@ -888,6 +888,11 @@ protected function save_game(BMGame $game) {
         // force game to proceed to the latest possible before saving
         $game->proceed_to_next_user_action();
 
+        // start transaction
+        if (!$this->isTest) {
```

Comment:
Remind me why we shouldn't do this for test invocations?  I thought the kinds of tests that came through apiResponder used a real database --- so why shouldn't they have transactions too?

Reply:
The simple answer is that things failed when I didn't have this conditional in place.

Unfortunately, I have no insight into why this is, I probably don't understand the test responder correctly.

Maybe you understand better?

Reply:
I don't, but what i do understand is that the `isTest` flag should do literally one thing: tell the code to use the buttonmen_test database rather than the buttonmen database.  I don't see any reason any of this stuff should be magical in responderTest, and the symptoms here are very weird.  Here i'm running a single responderTest, `test_interface_game_051`, which does a very normal `create_game` with Dead Guy and Gawaine --- nothing should be weird about this.

It fails like so:

```
Caught exception in BMInterface::save_game: There is no active transaction
Caught exception in BMInterface::create_game: There is no active transaction
F                                                                  2 / 2 (100%)

Time: 1.52 seconds, Memory: 10.00Mb

There was 1 failure:

1) responder03Test::test_interface_game_051
API call should succeed:
ARGS: array (
  'type' => 'createGame',
  'playerInfoArray' => 
  array (
    0 => 
    array (
      0 => 'responder003',
      1 => 'Dead Guy',
    ),
    1 => 
    array (
      0 => 'responder004',
      1 => 'Gawaine',
    ),
  ),
  'maxWins' => 3,
  'description' => '',
)
RETURN: array (
  'data' => NULL,
  'message' => 'Game create failed: There is no active transaction',
  'status' => 'failed',
)
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'ok'
+'failed'

/buttonmen/test/src/api/responderTestFramework.php:986
/buttonmen/test/src/api/responderTestFramework.php:1022
/buttonmen/test/src/api/responder03Test.php:3004

FAILURES!
Tests: 2, Assertions: 28, Failures: 1.
```

I annotated BMInterface with print statements, so i could see whether the transaction never gets started or whether we fall out of it or what --- in these, if the line ends with 1, `self::$conn->inTransaction()` succeeded; if not, it failed:

```
.entering create_game try loop
create_game: insert_new_game() happened
create_game: added players
create_game: set random buttons and custom recipes
create_game: loaded game
create_game: about to save game
in save_game()
save_game: done with proceed_to_next_user_action
save_game: done with beginTransaction: 1
save_game: done with resolve_random_button_selection: 1
save_game: done with save_basic_game_parameters: 1
save_game: done with save_button_recipes: 1
save_game: done with save_random_button_choice: 1
save_game: done with save_round_scores: 1
save_game: done with clear_swing_values_from_database: 1
save_game: done with clear_option_values_from_database: 1
save_game: done with save_swing_values_from_last_round: 1
save_game: done with save_swing_values_from_this_round: 1
save_game: done with save_option_values_from_last_round: 1
save_game: done with save_option_values_from_this_round: 1
save_game: done with save_player_game_decisions: 1
save_game: done with save_player_with_initiative: 1
save_game: done with save_players_awaiting_action: 1
save_game: done with regenerate_essential_die_flags: 1
save_game: done with mark_existing_dice_as_deleted: 1
save_game: done with save_active_dice: 1
save_game: done with save_captured_dice: 1
save_game: done with save_turbo_cache: 1
save_game: done with save_out_of_play_dice: 1
save_game: done with delete_dice_marked_as_deleted: 1
save_game: done with save_action_log: 
save_game: done with save_chat_log: 
save_game: about to commit: 
```

So we exited the transaction while calling `save_action_log()`?  Why?  I don't have an immediate answer, but i definitely think we need to know --- the *point* of unit tests is to catch all the weird things that might happen as we call functions in different ways --- RandomAI literally exercises BMGame logic.  It doesn't help us with the interface at all.

Even if this is a bug in a test harness, committing this is too much of a divergence between test behavior and live behavior --- it'll make our tests much more useless for catching any future problems with how we use the DB.  So i think we need to figure it out.

Reply:
Okay, i've learned a little more by instrumenting the part where it creates new PDO connections:

```
About to create new test PDO connection
entering create_game try loop
create_game: insert_new_game() happened
create_game: added players
create_game: set random buttons and custom recipes
create_game: loaded game
create_game: about to save game
in save_game()
save_game: done with proceed_to_next_user_action
save_game: done with beginTransaction: 1
save_game: done with resolve_random_button_selection: (1); (2)
save_game: done with save_basic_game_parameters: (1); (2)
save_game: done with save_button_recipes: (1); (2)
save_game: done with save_random_button_choice: (1); (2)
save_game: done with save_round_scores: (1); (2)
save_game: done with clear_swing_values_from_database: (1); (2)
save_game: done with clear_option_values_from_database: (1); (2)
save_game: done with save_swing_values_from_last_round: (1); (2)
save_game: done with save_swing_values_from_this_round: (1); (2)
save_game: done with save_option_values_from_last_round: (1); (2)
save_game: done with save_option_values_from_this_round: (1); (2)
save_game: done with save_player_game_decisions: (1); (2)
save_game: done with save_player_with_initiative: (1); (2)
save_game: done with save_players_awaiting_action: (1); (2)
save_game: done with regenerate_essential_die_flags: (1); (2)
save_game: done with mark_existing_dice_as_deleted: (1); (2)
save_game: done with save_active_dice: (1); (2)
save_game: done with save_captured_dice: (1); (2)
save_game: done with save_turbo_cache: (1); (2)
save_game: done with save_out_of_play_dice: (1); (2)
save_game: done with delete_dice_marked_as_deleted: (1); (2)
About to create new test PDO connection
in save_action_log(): (); (2)
after load_message_from_game_actions(): (); (2)
after log_game_actions(): (); (2)
save_game: done with save_action_log: (); (2)
About to create new test PDO connection
save_game: done with save_chat_log: (); (2)
save_game: about to commit: (); (2)
Caught exception in BMInterface::save_game: There is no active transaction
Caught exception in BMInterface::create_game: There is no active transaction
```

So that explains it --- `save_action_log()` and `save_chat_log()` each lead to a new conn being created.  That's clearly not going to end well.

Reply:
Here you go --- try this diff:

```diff
diff --git a/src/engine/BMInterface.php b/src/engine/BMInterface.php
index 116f746..a4fe0a5 100644
--- a/src/engine/BMInterface.php
+++ b/src/engine/BMInterface.php
@@ -83,8 +83,10 @@ class BMInterface {
         } else {
             require_once __DIR__.'/../database/mysql.inc.php';
         }
-        self::$conn = conn();
-        self::$db = new BMDB(self::$conn);
+        if (!(isset(self::$conn))) {
+            self::$conn = conn();
+            self::$db = new BMDB(self::$conn);
+        }
     }
 
     /**
```

If you make that change, all the responderTest tests pass.  There are still a bunch of failures in non-responder tests --- i'm out of time right now, but let's look into those.

Reply:
Right, that's a great start.

Reply:
I've checked with a similar instrumentation to you, and the issue with the BMInterface tests is now something different. The transactions are apparently completing successfully. I'll keep digging.

Reply:
I reckon it might be a timing issue, somehow.

I'm looking at the first error, which for me occurs in BMInterfaceForumTest->test_create_load_forum_posts().

The first line is

    $overview = $this->object->load_forum_overview(self::$userId1WithoutAutopass);

This returns with

    array(2) {
      ["boards"]=>
      array(0) {
      }
      ["timestamp"]=>
      int(1629510278)
    }

So, somehow, the boards aren't there to load. I don't know exactly what preparation goes on before this, but it looks like the test database is empty.

Reply:
Also, I checked, and self::$userId1WithoutAutopass looks like it has a valid value, so it's not a failure in the creation of the users.

Reply:
Okay, making progress. The PHP error log shows this error:

    Caught exception in BMInterface::save_game: DB select_single_value found no result

Reply:
I think that part of the issue is that self::$conn is an object of class TestDummyPDOConn. I'll try adding function definitions for the transaction-related functions, see if that fixes things.

Reply:
No, that was a confounding factor (that still needed fixing).

Reply:
The problem is that under test conditions, the BMInterface* objects are ending up as TestDummyPDOConn instead of PDO, because they're hitting the !isset(self::$conn) condition and then failing to create a proper PDO connection.

Reply:
The BMDB::$conn has class TestDummyPDOConn instead of PDO, which is causing the problem.

Reply:
And fixed! Just had to shift the condition brackets from Chaos's suggestion, and it all works. Pushing now.